ludwig/datasets/base_dataset.py

```diff
@@ -0,0 +1,118 @@
+#! /usr/bin/env python
+# coding=utf-8
+# Copyright (c) 2019 Uber Technologies, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import os
+import yaml
+import abc
+from pathlib import Path
+import pandas as pd
+
+
+"""A base class that defines the public interface for the ludwig dataset API.
+This includes the download, transform and converting the final transformed API
+into a resultant dataframe"""
+
+# define a default location for the cache
+DEFAULT_CACHE_LOCATION = str(Path.home().joinpath('.ludwig_cache'))
+
+
+class BaseDataset:
+
+    def __init__(self, dataset_name, cache_location):
+        self._dataset_name = dataset_name
+        self._initial_path = os.path.abspath(os.path.dirname(__file__))
+        self._config_file_location = os.path.join(self._initial_path, "./config/dataset_config.yaml")
+        with open(self._config_file_location) as config_file:
+            self._config_file_contents = yaml.load(config_file, Loader=yaml.FullLoader)
+        if cache_location is not None:
+            self._cache_location = cache_location
+        else:
+            self._cache_location = DEFAULT_CACHE_LOCATION
+        self._dataset_version = self._config_file_contents[dataset_name]["version"]
+        self._download_url = self._config_file_contents[dataset_name]["download_url"]
+        self._dataset_file_name = self._config_file_contents[dataset_name]["extracted_file_name"]
+        self._download_dir = Path.home().joinpath('.ludwig_cache').joinpath(self._dataset_name + "_"
```

Comment:
Replace with `self._cache_location`:

```
self._download_dir = os.path.join(self._cache_location, f'{self._dataset_name}_{self._dataset_version}')
```

Same for other file names below.

Also, I would consider moving the `_raw_file_name` into the Download mixin, processed filename into Process mixin, etc.

Reply:
Changed self._download_dir and moved mixin variables inside the mixins